### PR TITLE
Fix discretization symbolic one hot

### LINKED ISSUE #22044: Discretization layer with output_mode='one_hot' produces incorrect output shape for symbolic tensors
The `Discretization` layer with `output_mode='one_hot'` fails to compute the correct output shape when applied to a symbolic Keras `Input` tensor. It works correctly for eager tensors (like NumPy arrays) but does not add the expected one-hot dimension to the shape of a symbolic tensor.

**Code to Reproduce**

```python
from keras.layers import Discretization, Input
import keras
import numpy as np

# Eager tensor input
x = np.ones((2, 3, 4))
# Symbolic tensor input
i = Input((3, 4))

layer = Discretization(bin_boundaries=[1.0, 2.0, 3.0], output_mode='one_hot')

print(f"Keras version: {keras.__version__}")
print(f"Eager input shape: {x.shape}, Output shape: {layer(x).shape}")
print(f"Symbolic input shape: {i.shape}, Output shape: {layer(i).shape}")
```

**Actual Behavior**

```
Keras version: 3.13.1
Eager input shape: (2, 3, 4), Output shape: (2, 3, 4, 4)
Symbolic input shape: (None, 3, 4), Output shape: (None, 3, 4)
```

**Expected Behavior**

According to the documentation, the output for the symbolic tensor should have an additional dimension for the one-hot encoding. The expected output shape should be `(None, 3, 4, 4)`.

```
Keras version: 3.13.1
Eager input shape: (2, 3, 4), Output shape: (2, 3, 4, 4)
Symbolic input shape: (None, 3, 4), Output shape: (None, 3, 4, 4)
```

### PR DESCRIPTION
**Pull request by bittoby**
> # Fix Discretization layer symbolic tensor shape computation for one_hot mode
>
> ## **Before the fix:**
> * Eager tensor: (2, 3, 4) → (2, 3, 4, 5) ✅
> * Symbolic tensor: (None, 3, 4) → (None, 3, 4) ❌ (missing one-hot dimension)
> 
> ## **After the fix:**
> * Eager tensor: (2, 3, 4) → (2, 3, 4, 5) ✅
> * Symbolic tensor: (None, 3, 4) → (None, 3, 4, 5) ✅
> 
> ## **Solution**
> * Added proper shape computation logic by implementing:
> * compute_output_shape method: Handles different output modes correctly
> * Updated compute_output_spec method: Now uses the new shape computation
> 
> ## **Verification**
> The fix resolves the exact issue described in #22044:
> 
> ## ### Create layer
``` python
layer = keras.layers.Discretization(
    bin_boundaries=[0.0, 1.0, 2.0, 3.0], 
    output_mode='one_hot'
)
```

> ##  Test symbolic tensor (this was broken before)
``` python
symbolic_input = keras.Input(shape=(3, 4))
symbolic_output = layer(symbolic_input)
print(symbolic_output.shape)  # Now correctly shows (None, 3, 4, 5)
```
> Closes #22044

### CODE REVIEW

File: keras/src/layers/preprocessing/discretization.py [hunk 1]
```diff
@@ -213,8 +213,38 @@ def reset_state(self):
...
+            # If last dimension is 1, replace it with depth, otherwise append
+            if input_shape and input_shape[-1] == 1 and len(input_shape) > 1:
+                return tuple(input_shape[:-1]) + (depth,)
+            else:
+                return tuple(input_shape) + (depth,)
+        else:
+            # For multi_hot and count modes, the behavior is more complex
+            # and depends on the backend implementation. For now, we'll use
+            # a simplified approach that works for the most common cases.
+            # This may need refinement based on actual backend behavior.
```
Comment (hertschuh) [inline keras/src/layers/preprocessing/discretization.py]: How is it backend dependent? Can you provide examples?

Replacing the last dimension with "depth" should be correct always.

File: keras/src/layers/preprocessing/discretization_test.py [hunk 2]
```diff
@@ -231,3 +231,108 @@ def test_model_call_vs_predict_consistency(self):
...
+
+        This test ensures that issue #22044 is fixed - Discretization layer
+        with output_mode='one_hot' should produce correct output shape for
+        symbolic tensors.
+        """
+        # Test one_hot mode - should add depth dimension
+        layer_one_hot = layers.Discretization(
+            bin_boundaries=[0.0, 1.0, 2.0], output_mode="one_hot"
+        )
+        symbolic_input = layers.Input(shape=(3, 4))
```
Comment (hertschuh) [inline keras/src/layers/preprocessing/discretization_test.py]: Don't use `4` as the value for the last dimension. There are `4` bins, so won't be able to tell if it has actually been replaced with the "depth" value.

Just in general, using unique dimension values is a good approach.

File: keras/src/layers/preprocessing/discretization_test.py [hunk 3]
```diff
@@ -231,3 +231,108 @@ def test_model_call_vs_predict_consistency(self):
...
+        )
+        symbolic_output_int = layer_int(symbolic_input)
+        # int mode keeps same shape, so output should be (None, 3, 4)
+        self.assertEqual(symbolic_output_int.shape, (None, 3, 4))
+
+        # Test consistency between eager and symbolic tensors for one_hot mode
+        eager_input = np.random.uniform(0, 3, size=(2, 3, 4))
+        eager_output = layer_one_hot(eager_input)
+        # Shapes should be consistent (ignoring batch dimension)
+        self.assertEqual(eager_output.shape[1:], symbolic_output.shape[1:])
```
Comment (hertschuh) [inline keras/src/layers/preprocessing/discretization_test.py]: Why do you exclude the batch size in your comparison? The batch size better not change.

File: keras/src/layers/preprocessing/discretization_test.py [hunk 4]
```diff
@@ -231,3 +231,108 @@ def test_model_call_vs_predict_consistency(self):
...
+
+    def test_compute_output_shape_edge_cases(self):
+        """Test edge cases in compute_output_shape to improve coverage."""
+
+        # Test with num_bins instead of bin_boundaries
+        layer_num_bins = layers.Discretization(
+            num_bins=5, output_mode="one_hot"
+        )
```
Comment (hertschuh) [inline keras/src/layers/preprocessing/discretization_test.py]: Using `num_bins` hardly counts as and "edge case". Create a dedicated test called `test_compute_output_shape_num_bins`.

File: keras/src/layers/preprocessing/discretization_test.py [hunk 5]
```diff
@@ -231,3 +231,108 @@ def test_model_call_vs_predict_consistency(self):
...
+            ("multi_hot", (None, 3, 4), (None, 3, 3)),  # multi_hot - replace
+            ("count", (None, 3, 4), (None, 3, 3)),  # count - replace
+        ]
+
+        for mode, input_shape, expected_shape in modes_and_shapes:
+            layer = layers.Discretization(
+                bin_boundaries=[0.0, 1.0], output_mode=mode
+            )
+            result_shape = layer.compute_output_shape(input_shape)
+            self.assertEqual(result_shape, expected_shape)
```
Comment (hertschuh) [inline keras/src/layers/preprocessing/discretization_test.py]: As far as I can tell, this covers exactly the same use cases as `test_symbolic_tensor_output_shape`.

You know you can use parameterized tests instead of a for loop:
https://github.com/keras-team/keras/blob/master/keras/src/layers/preprocessing/discretization_test.py#L41-L93

File: keras/src/testing/test_utils.py [hunk 6]
```diff
@@ -161,3 +161,67 @@ def value_to_str(value):
         tests = new_tests
 
     return tests
+
+
+def assert_symbolic_shape_consistency(
+    test_case, layer, input_shape, expected_output_shape=None
+):
```
Comment (hertschuh) [inline keras/src/testing/test_utils.py]: This is not used and I don't think we need this. Revert this file.

File: keras/src/layers/preprocessing/discretization.py (hunk 1 above)
Comment (Crystora) [reply keras/src/layers/preprocessing/discretization.py]: Yes, you are right. I just thought of TF fastpath. But as you mentioned, output_shape should always be like above.

File: keras/src/layers/preprocessing/discretization_test.py (hunk 2 above)
Comment (Crystora) [reply keras/src/layers/preprocessing/discretization_test.py]: Sure. I will fix it to unique dimension value. Thanks.

File: keras/src/layers/preprocessing/discretization_test.py (hunk 3 above)
Comment (Crystora) [reply keras/src/layers/preprocessing/discretization_test.py]: You're absolutely right to question this! I just simply thought. The comment "ignoring batch dimension" is misleading and potentially incorrect. Let me compare the two batch sizes.  Symbolic should be None while eager should be an integer.

File: keras/src/layers/preprocessing/discretization_test.py (hunk 4 above)
Comment (Crystora) [reply keras/src/layers/preprocessing/discretization_test.py]: Yes.

File: keras/src/testing/test_utils.py (hunk 6 above)
Comment (Crystora) [reply keras/src/testing/test_utils.py]: Sure

File: keras/src/layers/preprocessing/discretization_test.py (hunk 5 above)
Comment (Crystora) [reply keras/src/layers/preprocessing/discretization_test.py]: Make sense.

**Comment by Crystora** [comment]
@hertschuh  Thanks for your great advice. Plz review again.

File: keras/src/layers/preprocessing/discretization.py (hunk 1 above)
Comment (hertschuh) [reply keras/src/layers/preprocessing/discretization.py]: Please remove this comment. It will cause confusion if somebody ever works on it.

File: keras/src/layers/preprocessing/discretization_test.py [hunk 7]
```diff
@@ -231,3 +231,293 @@ def test_model_call_vs_predict_consistency(self):
             backend.standardize_dtype(predict_output.dtype),
         )
         self.assertTrue(backend.is_int_dtype(model_call_output.dtype))
+
+    def test_symbolic_tensor_output_shape(self):
```
Comment (hertschuh) [inline keras/src/layers/preprocessing/discretization_test.py]: Please make this parameterized exactly like `test_compute_output_shape_modes` below.

File: keras/src/layers/preprocessing/discretization_test.py [hunk 8]
```diff
@@ -231,3 +231,293 @@ def test_model_call_vs_predict_consistency(self):
...
+
+        # Verify non-batch dimensions are identical
+        self.assertEqual(eager_output.shape[1:], symbolic_output.shape[1:])
+
+        # Verify total number of dimensions is the same
+        self.assertEqual(len(eager_output.shape), len(symbolic_output.shape))
+
+    @parameterized.named_parameters(
+        named_product(
+            [
```
Comment (hertschuh) [inline keras/src/layers/preprocessing/discretization_test.py:238]: You don't need `named_product` here since you're specifying the `testcase_name`s. Just do:

```
@parameterized.named_parameters(
    [
        {
             "testcase_name": "int_mode",
             ...
```

File: keras/src/layers/preprocessing/discretization_test.py [hunk 9]
```diff
@@ -231,3 +231,293 @@ def test_model_call_vs_predict_consistency(self):
...
+        layer = layers.Discretization(
+            bin_boundaries=[0.0, 1.0, 2.0], output_mode="one_hot"
+        )
+
+        # Test different batch sizes
+        batch_sizes = [1, 3, 5, 10]
+        input_shape = (4, 3)
+
+        for batch_size in batch_sizes:
+            with self.subTest(batch_size=batch_size):
```
Comment (hertschuh) [inline keras/src/layers/preprocessing/discretization_test.py]: This works, but use a parameterized test instead. This is where `named_product` helps:

```python
    @parameterized.named_parameters(named_product(batch_size=[1, 3, 5, 10]))
    def test_batch_dimension_consistency(self, batch_size):
```

File: keras/src/layers/preprocessing/discretization_test.py [hunk 10]
```diff
@@ -231,3 +231,293 @@ def test_model_call_vs_predict_consistency(self):
...
+                expected_shape = input_shape + (4,)  # 3 boundaries = 4 bins
+                self.assertEqual(eager_output.shape[1:], expected_shape)
+                self.assertEqual(symbolic_output.shape[1:], expected_shape)
+
+    def test_compute_output_shape_num_bins(self):
+        """Test compute_output_shape with num_bins parameter."""
+
+        # Test different num_bins values with different output modes
+        test_cases = [
+            {
```
Comment (hertschuh) [inline keras/src/layers/preprocessing/discretization_test.py]: Make this a proper parameterized test.

File: keras/src/layers/preprocessing/discretization_test.py [hunk 11]
```diff
@@ -231,3 +231,293 @@ def test_model_call_vs_predict_consistency(self):
...
+                    case["expected_shape"],
+                    f"Failed for num_bins={case['num_bins']}, "
+                    f"mode={case['output_mode']}",
+                )
+
+        # Test various num_bins values with one_hot mode
+        num_bins_values = [2, 3, 5, 10, 20, 100]
+        input_shape = (None, 4, 3)
+
+        for num_bins in num_bins_values:
```
Comment (hertschuh) [inline keras/src/layers/preprocessing/discretization_test.py]: Make this a parameterized test, follow the batch_size pattern for this one.

File: keras/src/layers/preprocessing/discretization_test.py [hunk 12]
```diff
@@ -231,3 +231,293 @@ def test_model_call_vs_predict_consistency(self):
...
+                )
+
+                result_shape = layer.compute_output_shape(input_shape)
+                expected_shape = (None, 4, 3, num_bins)
+
+                self.assertEqual(
+                    result_shape,
+                    expected_shape,
+                    f"Failed for num_bins={num_bins}: expected "
+                    f"{expected_shape}, got {result_shape}",
```
Comment (hertschuh) [inline keras/src/layers/preprocessing/discretization_test.py]: No need for this, all this information will already be displayed.

File: keras/src/layers/preprocessing/discretization_test.py [hunk 13]
```diff
@@ -231,3 +231,293 @@ def test_model_call_vs_predict_consistency(self):
...
+                )
+
+        # Test edge case: num_bins=1
+        layer_single_bin = layers.Discretization(
+            num_bins=1, output_mode="one_hot"
+        )
+        result = layer_single_bin.compute_output_shape((None, 3, 2))
+        expected = (None, 3, 2, 1)
+        self.assertEqual(result, expected)
```
Comment (hertschuh) [inline keras/src/layers/preprocessing/discretization_test.py]: No need for special code for this, just add `1` to the list of `num_bins` above.

File: keras/src/layers/preprocessing/discretization_test.py [hunk 14]
```diff
@@ -231,3 +231,293 @@ def test_model_call_vs_predict_consistency(self):
...
+            with self.subTest(input_shape=input_shape):
+                result_shape = layer.compute_output_shape(input_shape)
+                expected_shape = input_shape + (7,)  # Append depth for one_hot
+
+                self.assertEqual(
+                    result_shape,
+                    expected_shape,
+                    f"Failed for input shape {input_shape}: expected "
+                    f"{expected_shape}, got {result_shape}",
+                )
```
Comment (hertschuh) [inline keras/src/layers/preprocessing/discretization_test.py]: I don't think this is needed, you got `compute_output_shape` well covered.

**Comment by Crystora** [comment]
@hertschuh  I fixed problems. Plz review again. Thanks.

**Comment by Crystora** [comment]
@hertschuh I already fixed all what you mentioned. Plz review again.

File: keras/src/layers/preprocessing/discretization_test.py [hunk 15]
```diff
@@ -231,3 +232,268 @@ def test_model_call_vs_predict_consistency(self):
...
+
+        # Create symbolic input and get output
+        symbolic_input = layers.Input(shape=input_shape)
+        symbolic_output = layer(symbolic_input)
+
+        # Verify symbolic output shape
+        self.assertEqual(symbolic_output.shape, expected_shape)
+
+        # Test consistency between eager and symbolic tensors for one_hot mode
+        if output_mode == "one_hot":
```
Comment (hertschuh) [inline keras/src/layers/preprocessing/discretization_test.py]: Why not for all? There's nothing special about "one_hot", you can do it for all.

File: keras/src/layers/preprocessing/discretization_test.py [hunk 16]
```diff
@@ -231,3 +232,268 @@ def test_model_call_vs_predict_consistency(self):
...
+    def _test_num_bins_one_hot_mode(self, num_bins):
+        """Test various num_bins values with one_hot mode."""
+        input_shape = (None, 4, 3)
+
+        layer = layers.Discretization(num_bins=num_bins, output_mode="one_hot")
+
+        result_shape = layer.compute_output_shape(input_shape)
+        expected_shape = (None, 4, 3, num_bins)
+
+        self.assertEqual(result_shape, expected_shape)
```
Comment (hertschuh) [inline keras/src/layers/preprocessing/discretization_test.py]: This is already covered by `one_hot_mode_4_bins` line 434, you can remove.

Comment (Crystora) [reply keras/src/layers/preprocessing/discretization_test.py]: Thanks, I will fix it.

File: keras/src/layers/preprocessing/discretization_test.py (hunk 15 above)
Comment (Crystora) [reply keras/src/layers/preprocessing/discretization_test.py]: Yes, there should be all of the modes. Thanks.

**Comment by Crystora** [comment]
#[hertschuh](https://github.com/hertschuh), thanks to your comments, I found another issue upon multi_hot and count mode and fix the problems. In the previous times,  since #22044 issue is about one_hot mode, I only focused on one_hot mode. The issues are about output_shape in multi_hot and count modes and these are slightly different from expectations.
```
layer = Discretization(
    bin_boundaries=[0.0, 1.0, 2.0, 5.0], output_mode="multi_hot"
)

input_shape = (3, 4)
symbolic_input = layers.Input(shape=input_shape)
symbolic_output = layer(symbolic_input)

Output: 
Eager: Input (2, 3, 4) -> Output torch.Size([2, 4, 5])

Expectation:
Eager: Input (2, 3, 4) -> Output torch.Size([2, 3, 5])

Symbolic:
Input (None, 3, 4) -> Output torch.Size([None, 4, 5])
```

I fix the problem, which symbolic tensor output_shape is same to original eager tensor output_shape.
Thanks for your kind attention. Plz review again.

**Review by hertschuh — APPROVED** [review]
Thanks!

**Comment by Crystora** [comment]
[hertschuh](https://github.com/hertschuh) Thank you.
